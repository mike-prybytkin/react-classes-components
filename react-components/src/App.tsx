import AboutUs from 'pages/AboutUs/AboutUs';
import Page404 from 'pages/404/404';
import React from 'react';
import { Route, Routes } from 'react-router-dom';
import './App.scss';

function App() {
  return (
    <Routes>
      <Route path="/about" element={<AboutUs />} />
      <Route path="/404" element={<Page404 />} />
    </Routes>
  );
}

export default App;
