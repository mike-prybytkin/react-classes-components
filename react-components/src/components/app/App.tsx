import React from 'react';
import { Route, Routes, Navigate } from 'react-router-dom';
import AboutUs from 'pages/about-us/about-us';
import Page404 from 'pages/404/404';
import Main from 'pages/main/main';
import Header from 'components/header/header';
import { IAppState } from './types';
import mockText from 'mocks/text';
import Form from 'pages/form/form';
import { IFetchData } from './types';

class App extends React.Component<unknown, IAppState> {
  constructor(props: unknown) {
    super(props);
    this.state = {
      data: [],
      isLoading: false,
      fetchError: '',
      querySearch: '',
    };
  }

  async fetchData() {
    const URL = `https://rickandmortyapi.com/api/character/?name=${this.state.querySearch}`;
    this.setState({ isLoading: true });
    try {
      const response = await fetch(URL);
      const fetchData: IFetchData = await response.json();
      this.setState({
        data: fetchData.results,
        isLoading: false,
      });
    } catch (error) {
      this.setState({
        fetchError: error,
      });
    }
  }

  onSearch = (querySearch: string) => {
    this.setState({
      querySearch: querySearch,
    });
  };

  componentDidMount() {
    this.fetchData();
  }

  componentDidUpdate(prevProps: Readonly<unknown>, prevState: Readonly<IAppState>): void {
    if (prevState.querySearch !== this.state.querySearch) {
      this.fetchData();
    }
  }

  render() {
    return (
      <div className="App" data-testid="app">
        <Header onSearch={this.onSearch} />
        <main className="main">
          <Routes>
            <Route
              path="/"
              element={<Main cards={this.state.data} heading={mockText.headingMain} />}
            />
            <Route
              path="/about"
              element={
                <AboutUs heading={mockText.headingAboutUs} message={mockText.pageNotReady} />
              }
            />
            <Route
              path="/form"
              element={
                <Form
                  heading={mockText.headingForm}
                  noCreatedUserMessage={mockText.notCreatedUserInForm}
                />
              }
            />
            <Route
              path="/notFound"
              element={
                <Page404
                  heading={mockText.heading404}
                  backToHomelinkText={mockText.linkBackToHome}
                />
              }
            />
            <Route path="/*" element={<Navigate to="/notFound" />} />
          </Routes>
          {this.state.isLoading && <p className="main__message_loading">{mockText.loading}</p>}
        </main>
      </div>
    );
  }
}

export default App;
