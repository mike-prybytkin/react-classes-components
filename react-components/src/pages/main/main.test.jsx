import React from 'react';
import { render, screen } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import App from '../../components/app/app';

describe('Main', () => {
  it('render main heading', () => {
    render(
      <BrowserRouter>
        <App />
      </BrowserRouter>
    );
    const mainHeading = screen.getByText(/Welcome to our store/i);

    expect(mainHeading).toBeInTheDocument();
  });
});
