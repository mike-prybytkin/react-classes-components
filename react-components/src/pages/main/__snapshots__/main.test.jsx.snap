// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main component should create snapshot 1`] = `
<DocumentFragment>
  <h1>
    Welcome to us!
  </h1>
  <div
    class="cards-wrapper"
  >
    <div
      class="card"
    >
      <img
        alt="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
        class="card__img"
        src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
      />
      <h3
        class="card__title"
      >
        Rick Sanchez
      </h3>
      <button
        class="card__button"
      >
        Learn more
      </button>
    </div>
    <div
      class="card"
    >
      <img
        alt="https://rickandmortyapi.com/api/character/avatar/2.jpeg"
        class="card__img"
        src="https://rickandmortyapi.com/api/character/avatar/2.jpeg"
      />
      <h3
        class="card__title"
      >
        Morty Smith
      </h3>
      <button
        class="card__button"
      >
        Learn more
      </button>
    </div>
  </div>
</DocumentFragment>
`;
